### ISSUE TITLE
Request for Data Preparation Scripts

### ISSUE THREAD

**Issue by wlin-at** [AtlasAnalyticsLab/2DMamba#16]
Dear Authors,

Thank you for your great work. I wanted to ask whether you plan to release the scripts for WSI data preparation (segmentation, patching, and feature extraction). I have been trying to follow the data processing pipelines provided in the [CLAM repo](https://github.com/mahmoodlab/CLAM), but I encountered a few dataset-specific issues:

Segmentation and filtering hyperparameters.  
It seems that the optimal hyperparameters can vary across datasets. For example, the CLAM repo provides presets for TCGA, BWH_biopsy, and BWH_resection, but not for others. I noticed that using unsuitable hyperparameters sometimes leads to stalled feature extraction jobs and zombie processes without raising explicit errors.

Thank you very much for your time and consideration.

Best,

Wei

**Comment by wlin-at** [comment]
I am following the two-step pipeline of segmentation/patching and feature extraction to compute UNI features for 1,040 TCGA-BRCA whole-slide images downloaded from GDC.

Commands used
Segmentation & Patching:
```
python create_patches_fp.py \
  --source /system/user/publicdata/WSI_data/TCGA_BRCA/TCGA_BRCA_1040 \
  --save_dir /system/user/publicdata/WSI_data/TCGA_BRCA/TCGA_BRCA_processed_w_template_new \
  --patch_size 512 \
  --step_size 512 \
  --preset tcga.csv \
  --seg --patch --stitch

```
Feature extraction
```
CUDA_VISIBLE_DEVICES=4 python extract_features_fp.py \
  --model_name uni_v1 \
  --data_h5_dir /system/user/publicdata/WSI_data/TCGA_BRCA/TCGA_BRCA_processed_w_template_new \
  --data_slide_dir /system/user/publicdata/WSI_data/TCGA_BRCA/TCGA_BRCA_1040 \
  --csv_path /system/user/publicdata/WSI_data/TCGA_BRCA/TCGA_BRCA_processed_w_template_new/process_list_autogen_updated.csv \
  --feat_dir /system/user/publicdata/WSI_data/TCGA_BRCA/TCGA_BRCA_features_w_template_new/ \
  --batch_size 512 \
  --slide_ext .svs

```
Setup
- Patch size / stride: 512
- Feature extractor: [UNI encoder (v1)](https://huggingface.co/MahmoodLab/UNI)
- GPU: NVIDIA A40 (46 GB VRAM)
- Batch size: 512 (uses ~8 GB VRAM only)

Observations

- Segmentation and patching run smoothly.
- During feature extraction, the job often stalls after several hours.
- No error is thrown; the process just hangs indefinitely (appears like a zombie process).
- This occurs repeatedly, requiring me to manually kill the process, update the process list, and restart.
- The slides where this happens are not unusually large (e.g., ~16k patches, close to average).

Question
Have you encountered similar hanging behavior when running UNI feature extraction?
Is there a recommended workaround (e.g., smaller batch size, different dataloader settings, or specific environment configurations) to avoid these stalls?

Thank you in advance for your help!

Here is the distribution of number of patches (with patch size/stride of 512) in TCGA-BRCA (1040 WSI)

<img width="587" height="447" alt="Image" src="https://github.com/user-attachments/assets/93222bea-cb41-4bc9-adde-8671e3857c9e" />

**Comment by wlin-at** [comment]
When processing WSI datasets (e.g., BRACS), I noticed that the default segmentation and filter parameters sometimes result in zero valid tissue contours being detected. In these cases, the function WSI_object.process_contours(**kwargs) does not generate an .h5 patch file for the corresponding slide.

This leads to situations where:
- A mask .jpg file is still created (from WSI_object.visWSI()),
- But no .h5 file is saved in the patches folder,
- Consequently, no stitch .jpg file is produced either (since stitching requires the .h5 file).

Could you share the hyperparameters used for segmentation and patching for these WSI datasets?

Thanks in advance!
